### PR TITLE
Add suport for Python 3.11

### PR DESCRIPTION
**Pull request by maxrjones**
**Description of proposed changes**

Run tests on Python 3.11, released last October.

Waiting on the next TensorFlow release

Fixes #

### CODE REVIEW

**Comment by weiji14** [comment]
Tracking upstream tensorflow Python 3.11 support at https://github.com/tensorflow/tensorflow/issues/58032. [Tensorflow 2.12.0-rc0](https://github.com/tensorflow/tensorflow/releases/tag/v2.12.0-rc0) was just released on 15 Feb 2023, so hopefully just another month or two to wait :crossed_fingers:

**Comment by maxrjones** [comment]
> Tracking upstream tensorflow Python 3.11 support at https://github.com/tensorflow/tensorflow/issues/58032. [Tensorflow 2.12.0-rc0](https://github.com/tensorflow/tensorflow/releases/tag/v2.12.0-rc0) was just released on 15 Feb 2023, so hopefully just another month or two to wait 🤞

https://github.com/tensorflow/tensorflow/releases/tag/v2.12.0 was just released, so testing this out again now

**Comment by maxrjones** [comment]
The failing CI for dev branches relates to https://github.com/dask/dask/pull/10070. We could either remove the `--no-deps` flag for the pip upgrade command or wait for the next Dask release (scheduled for tomorrow).

File: .github/workflows/main.yaml
Comment (weiji14) [inline .github/workflows/main.yaml:1]: Did you also want to add Python 3.11 to the test-upstream job at L53? Even though that job is failing on Python 3.10 for some reason.

Comment (weiji14) [reply .github/workflows/main.yaml:1]: > The failing CI for dev branches relates to [dask/dask#10070](https://github.com/dask/dask/pull/10070). We could either remove the `--no-deps` flag for the pip upgrade command or wait for the next Dask release (scheduled for tomorrow).

I was looking at that, but it seems that we're installing from dask's main branch

https://github.com/xarray-contrib/xbatcher/blob/a95febe83fd961d60e8647aaf44e2570d2ef71d1/.github/workflows/main.yaml#L63-L64

So a new dask release probably won't work. Maybe remove `--no-deps` instead?

Comment (maxrjones) [reply .github/workflows/main.yaml:1]: Good recommendation - I'll make that change. see https://github.com/xarray-contrib/xbatcher/pull/148#issuecomment-1481826323 for an explanation of the failure

**Review by weiji14 — APPROVED** [review]

**Pull request merged**